p2p/profile_service.go

```diff
@@ -0,0 +1,227 @@
+package p2p
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/helpers"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	protocol "github.com/libp2p/go-libp2p-core/protocol"
+	"github.com/qri-io/qri/config"
+	"github.com/qri-io/qri/event"
+	"github.com/qri-io/qri/repo"
+	"github.com/qri-io/qri/repo/profile"
+)
+
+const (
+	// ProfileProtocolID is the protocol id for the profile exchange service
+	ProfileProtocolID = protocol.ID("/qri/profile/0.1.0")
+	// ProfileTimeout is the length of time we will wait for a response in a
+	// profile exchange
+	ProfileTimeout = time.Minute * 2
+)
+
+// ProfileService manages the profile exchange. This exchange should happen
+// whenever a node connects to a new peer
+type ProfileService struct {
+	host    host.Host
+	repo    repo.Repo
+	pub     event.Publisher
+	peersMu *sync.Mutex // lock for the peers map
+	peers   map[peer.ID]chan struct{}
+}
+
+// NewQriProfileService creates an profile exchange service and adds
+// the `ProfileHandler` to the host
+func NewQriProfileService(r repo.Repo, p event.Publisher) *ProfileService {
+	ps := &ProfileService{
+		repo:    r,
+		pub:     p,
+		peersMu: &sync.Mutex{},
+		peers:   map[peer.ID]chan struct{}{},
+	}
+	// h.SetStreamHandler(ProfileProtocolID, ps.ProfileHandler)
+	return ps
+}
+
+// IsQriPeer returns true if the given peer ID is a peer we
+// have connected to this session
+func (ps *ProfileService) IsQriPeer(pid peer.ID) bool {
+	ps.peersMu.Lock()
+	_, ok := ps.peers[pid]
+	ps.peersMu.Unlock()
+	return ok
+}
+
+// StartProfileService adds a host and a profile handler to the host
+// to the profile service
+func (ps *ProfileService) StartProfileService(h host.Host) {
+	ps.host = h
+	h.SetStreamHandler(ProfileProtocolID, ps.ProfileHandler)
+}
+
+// ProfileHandler listens for profile requests
+// it sends it's node's profile on the given stream
+// whenever a request comes in
+func (ps *ProfileService) ProfileHandler(s network.Stream) {
+
+	var (
+		err error
+		pro *profile.Profile
+	)
+
+	p := s.Conn().RemotePeer()
+
+	defer func() {
+		helpers.FullClose(s)
```

Comment:
Mind adding your clarification from review as a comment here? I find the `FullClose` method name confusing. (`FullCloseStream` would have been nicer)

```go
// close the stream, and wait for the other end of the stream to close as well
// this won't close the underlying connection
helpers.FullClose(s)
````

Adding explanatory comments will help others understand our p2p code better. six-months-from-now us will be thankful